File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.cpp
class Solution {
public:
    vector<vector<bool>> a;
    vector<vector<bool>> p;
    int m,n;
    vector<vector<int>> ans;
    void dfs(vector<vector<int>> &h ,vector<vector<bool>> &visited,int i,int j)
    {
        if(visited[i][j]) return;
        
        visited[i][j]=true;
        
        if(a[i][j] && p[i][j]) ans.push_back({i,j});
        
        if(i>0   && h[i-1][j]>= h[i][j]) dfs(h,visited,i-1,j);
        if(j>0   && h[i][j-1]>= h[i][j]) dfs(h,visited,i,j-1);
        if(i<m-1 && h[i+1][j]>= h[i][j]) dfs(h,visited,i+1,j);
        if(j<n-1 && h[i][j+1]>= h[i][j]) dfs(h,visited,i,j+1);
    }
   
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& h) {
        m=h.size(),n=h[0].size();
        a=p=vector(m,vector(n,false));
        for(int i=0;i<n;i++) dfs(h,p,0,i),dfs(h,a,m-1,i);
        for(int i=0;i<m;i++) dfs(h,p,i,0),dfs(h,a,i,n-1);
        
        // for(int i=0;i<m;i++) 
        //     for(int j=0;j<n;j++) 
        //         if(a[i][j] && p[i][j]) ans.push_back({i,j});
        
        return ans;
    }
};

File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.java
class Solution {
    
    boolean[][] pacific;
    boolean[][] atlantic;
    int m,n;
    List<List<Integer>> result = new ArrayList<List<Integer>>();
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        m = heights.length; n = heights[0].length;
        pacific  = new boolean[m][n];
        atlantic = new boolean[m][n];
        for(int i=0;i<n;i++) 
        {
            dfs(0,i,heights,pacific);dfs(m-1,i,heights,atlantic);
        }
        for(int i=0;i<m;i++)
        {
            dfs(i,0,heights,pacific);dfs(i,n-1,heights,atlantic);
        }
        return result;
    }
    private void dfs(int i, int j,int[][] heights,boolean[][] visited){
        if(visited[i][j]) return;
        
        visited[i][j]=true;
        
        if(atlantic[i][j] && pacific[i][j]) result.add(Arrays.asList(i,j));
        if(i>0   && heights[i-1][j]>= heights[i][j]) dfs(i-1,j,heights,visited);
        if(j>0   && heights[i][j-1]>= heights[i][j]) dfs(i,j-1,heights,visited);
        if(i<m-1 && heights[i+1][j]>= heights[i][j]) dfs(i+1,j,heights,visited);
        if(j<n-1 && heights[i][j+1]>= heights[i][j]) dfs(i,j+1,heights,visited);
    }
}